### PR TITLE
Implement stochastic gradient langevin dynamics

### LINKED ISSUE #131: Add SgLd
Starting from the SgMCMCJAX implementation.

### PR DESCRIPTION
**Pull request by rlouf**
With the gradient estimator in Robbins & Monro (1951). Closes #131.

### CODE REVIEW

File: blackjax/mcmc/sgld.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+"""Public API for the Stochastic gradient Langevin Dynamics kernel."""
+from typing import Callable, NamedTuple
+
+from blackjax.mcmc.diffusion import overdamped_langevin
+from blackjax.types import PRNGKey, PyTree
+
+__all__ = ["SgLDState", "SgLDInfo", "init", "kernel"]
+
+
+class SgLDState(NamedTuple):
```
Comment (jeremiecoullon) [inline blackjax/mcmc/sgld.py:10]: nitpick: the lowercase g looks odd to me. The standard way of writing the abreviation (in papers) is "SGLD", so `SGLDState` would feel more natural to me

**Comment by rlouf** [comment]
The gradient estimator and SGLD algorithms are implemented. I also re-implemented [SGMCMCJax's MLP example](https://sgmcmcjax.readthedocs.io/en/latest/nbs/BNN.html) locally. Left to do:
- Add tests for both SGLD and the gradient estimator;
- Add a notebook with the MLP example on MNIST.

File: blackjax/sgmcmc/gradients.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+) -> Callable:
+    """Builds a simple gradient estimator.
+
+    This estimator first appeared in [1]_.
+
+    Parameters
+    ----------
+    logprior_fn
+        The log-probability density function corresponding to the prior distribution.
+    loglikelihood_fn
```
Comment (rlouf) [inline blackjax/sgmcmc/gradients.py:20]: Here I should specify the expected signature for `loglikelihood_fn` and `logprior_fn` as it is a potential source of confusion.

File: tests/test_sampling.py [hunk 3]
```diff
@@ -110,21 +110,21 @@ def test_linear_regression(self, case, is_mass_matrix_diagonal):
...
     },
     {
         "algorithm": blackjax.nuts,
-        "initial_position": jnp.array(100.0),
-        "parameters": {"step_size": 0.1, "inverse_mass_matrix": jnp.array([0.1])},
+        "initial_position": jnp.array(3.0),
+        "parameters": {"step_size": 4.0, "inverse_mass_matrix": jnp.array([1.0])},
         "num_sampling_steps": 6000,
-        "burnin": 5_000,
+        "burnin": 1_000,
```
Comment (junpenglao) [inline tests/test_sampling.py:127]: Could you put these changes to a separate PR?

File: blackjax/kernels.py [hunk 4]
```diff
@@ -389,6 +391,59 @@ def step_fn(rng_key: PRNGKey, state):
...
+
+    def __new__(  # type: ignore[misc]
+        cls,
+        grad_estimator_fn: Callable,
+        schedule_fn: Callable,
+    ) -> SamplingAlgorithm:
+
+        step = cls.kernel(grad_estimator_fn)
+
+        def init_fn(position: PyTree, batch):
```
Comment (junpenglao) [inline blackjax/kernels.py:432]: suggestion: `batch` -> `data_batch` to emphasize this is batch of data

File: blackjax/sgmcmc/diffusion.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+def generate_gaussian_noise(rng_key: PRNGKey, position):
+    position_flat, unravel_fn = jax.flatten_util.ravel_pytree(position)
+    noise_flat = jax.random.normal(rng_key, shape=jnp.shape(position_flat))
+    return unravel_fn(noise_flat)
+
+
+def overdamped_langevin(logprob_grad_fn):
+    """Euler solver for overdamped Langevin diffusion."""
+
+    def one_step(rng_key, state: DiffusionState, step_size: float, batch: tuple = ()):
```
Comment (junpenglao) [inline blackjax/sgmcmc/diffusion.py:26]: It's better to type hint `batch` as pytree.

File: tests/test_sampling.py (hunk 3 above)
Comment (rlouf) [reply tests/test_sampling.py:127]: I feel like a separate commit is enough.

**Comment by rlouf** [comment]
I have a (small) dilemma for the API. Let's assume that we're doing some supervised learning, so the data consists of a design matrix `X` and result vector `y`. Let's call the parameters of the model `position`. The `logprior_fn` is only a function of `position`:

```python
def logprior_fn(position):
   pass
```

But `loglikelihood_fn` is a function of `position`, `X` and `y`. We have several choices for the API. I think that in either case the following signature for the step function is preferable:

```python
data_batch = (X, y)
new_state = sgld.step(rng_key, state, data_batch)
```

I am just not sure which of 

```python
def loglikelihood_fn(position, X, y):
   pass
```

or 

```python
def loglikelihood_fn(position, data):
   X, y = data
   pass
```

we should ask the user to provide. I'd say the latter (consistency), but I'd like your opinion nevertheless.
